internal/controllers/machine/machine_controller.go

```diff
@@ -488,6 +499,74 @@ func (r *Reconciler) isDeleteNodeAllowed(ctx context.Context, cluster *clusterv1
+// plane actual major.minor version.
+func (r *Reconciler) isVersionAllowed(ctx context.Context, machine *clusterv1.Machine, cluster *clusterv1.Cluster) (bool, error) {
+	// Skip the check if Version nil, because it is an optional field.
+	if machine.Spec.Version == nil {
+		return true, nil
+	}
+
+	// Skip the check if ControlPlaneRef is nil, because it is an optional field.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return true, nil
+	}
+
+	mv, err := semver.ParseTolerant(*machine.Spec.Version)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse machine version")
+	}
+
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Spec.ControlPlaneRef.Namespace)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to get control plane object")
+	}
+
+	controlPlaneActualVersionStr, err := contract.ControlPlane().StatusVersion().Get(controlPlane)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to read control plane actual version")
+	}
+	// controlPlaneVersionStr is not nil, because Get did not return an error.
+	controlPlaneActualVersion, err := semver.ParseTolerant(*controlPlaneActualVersionStr)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse control plane actual version")
+	}
+
+	controlPlaneDesiredVersionStr, err := contract.ControlPlane().Version().Get(controlPlane)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to read control plane desired version")
+	}
+	// controlPlaneVersionStr is not nil, because Get did not return an error.
+	controlPlaneDesiredVersion, err := semver.ParseTolerant(*controlPlaneDesiredVersionStr)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to parse control plane desired version")
+	}
+
+	if util.IsControlPlaneMachine(machine) {
+		if version.Compare(mv, controlPlaneDesiredVersion, version.IgnorePatchVersion()) == 1 {
+			return false, errors.Errorf("machine major.minor version (%d.%d) must be less than or equal to"+
+				"the control plane desired major.minor (%d.%d) version", mv.Major, mv.Minor,
+				controlPlaneDesiredVersion.Major, controlPlaneDesiredVersion.Minor)
+		}
+		if version.Compare(mv, controlPlaneActualVersion, version.IgnorePatchVersion()) == -1 {
+			return false, errors.Errorf("machine major.minor version (%d.%d) must be greater than or equal to"+
+				"the control plane actual major.minor (%d.%d) version", mv.Major, mv.Minor,
+				controlPlaneActualVersion.Major, controlPlaneActualVersion.Minor)
+		}
+	} else {
+		if version.Compare(mv, controlPlaneActualVersion, version.IgnorePatchVersion()) == 1 {
+			return false, errors.Errorf("machine major.minor version (%d.%d) must be less than or equal to"+
+				"the control plane actual major.minor (%d.%d) version", mv.Major, mv.Minor,
+				controlPlaneActualVersion.Major, controlPlaneActualVersion.Minor)
+		}
+	}
+
+	return true, nil
```

Comment:
Since gocritic's complaining anyway, what about

```suggestion
	if util.IsControlPlaneMachine(machine) {
		if version.Compare(mv, controlPlaneDesiredVersion, version.IgnorePatchVersion()) == 1 {
			return false, errors.Errorf("machine major.minor version (%d.%d) must be less than or equal to"+
				"the control plane desired major.minor (%d.%d) version", mv.Major, mv.Minor,
				controlPlaneDesiredVersion.Major, controlPlaneDesiredVersion.Minor)
		}
		if version.Compare(mv, controlPlaneActualVersion, version.IgnorePatchVersion()) == -1 {
			return false, errors.Errorf("machine major.minor version (%d.%d) must be greater than or equal to"+
				"the control plane actual major.minor (%d.%d) version", mv.Major, mv.Minor,
				controlPlaneActualVersion.Major, controlPlaneActualVersion.Minor)
		}
		return true, nil
	} 
	if version.Compare(mv, controlPlaneActualVersion, version.IgnorePatchVersion()) == 1 {
		return false, errors.Errorf("machine major.minor version (%d.%d) must be less than or equal to"+
			"the control plane actual major.minor (%d.%d) version", mv.Major, mv.Minor,
			controlPlaneActualVersion.Major, controlPlaneActualVersion.Minor)
	}
	return true, nil
```